packages/artillery-plugin-publish-metrics/lib/open-telemetry/index.js

```diff
@@ -395,6 +416,62 @@ class OTelReporter {
     }
   }
 
+  async runOtelTracingForPlaywright(
+    page,
+    vuContext,
+    events,
+    userProcessor,
+    specName
+  ) {
+    const playwrightTracer = trace.getTracer('artillery-playwright');
+
+    // Start a new active span for the scenario
+    return await playwrightTracer.startActiveSpan(
+      specName || 'Scenario execution',
+      { kind: SpanKind.CLIENT },
+      async (span) => {
+        if (this.traceConfig.attributes) {
+          span.setAttributes(this.traceConfig.attributes);
+        }
+        try {
+          // Attach traceStep function to the context
+          vuContext.funcs.traceStep = await this.traceStep(
+            span,
+            playwrightTracer,
+            this.traceConfig.attributes
+          );
+          // Execute the user-provided processor function within the context of the new span
+          await userProcessor(page, vuContext, events);
+        } catch (err) {
+          span.setStatus({
+            code: SpanStatusCode.ERROR,
+            message: err.message
+          });
+          debug(err);
+        } finally {
+          // End the scenario span
+          span.end();
+        }
+      }
+    );
+  }
+
+  // Allows users to wrap a span around a step or set of steps  (transaction) and add attributes to it. It also sends the span into the callback so users have ability to set additional attributes, events etc.
+  async traceStep(parent, tracer, configAttributes) {
+    return async function (name, callback, attributes) {
+      const ctx = trace.setSpan(context.active(), parent);
+      const span = tracer.startSpan(name, undefined, ctx);
+      if (configAttributes) {
+        span.setAttributes(configAttributes);
+      }
+      await callback(span);
+      if (attributes) {
+        span.setAttributes(attributes);
+      }
+      span.end();
+    };
+  }
```

Comment:
Hey @InesNi 👋 ! Nice work on this, it's great that this is API is compatible with the one introduced in the PR here https://github.com/artilleryio/artillery/pull/2173/files - it will make it easy for the user.

However, in the event that the user has already wrapped their test with steps using that method, they would now need to change their test to use `traceStep`. Ideally, they would just need to enable the publish-metrics plugin. 

I'm wondering if it might not be better to call this function `step` as well, and have it override my `step` function when tracing is enabled. And then you can have your function emit custom timers too. This would make the transition for the user transparent - they would just have the optional configuration available for them of setting `attributes` if desired.

Reply:
Oh I see, good thinking! 🙌🏻 I will change the name and have the function emit the timers then for now. I would also turn this PR into a draft so I can adjust the behaviour once your PR has been merged, if you think that makes sense @bernardobridge?